javatests/arcs/sdk/HandleHolderBaseTest.kt

```diff
@@ -0,0 +1,213 @@
+    val handleHolder = DummyHandleHolder()
+    handleHolder.setHandle("readHandle", mock())
+
+    assertFailsWith<IllegalArgumentException> {
+      handleHolder.setHandle("readHandle", mock())
+    }.also {
+      assertThat(it.message).isEqualTo("TestParticle.readHandle has already been initialized.")
+    }
+  }
+
+  @Test
+  fun getHandle_handleNameInSpec_returnsHandle() {
+    val handleHolder = DummyHandleHolder()
+    val mock = mock<Handle>()
+    handleHolder.setHandle("writeHandle", mock)
+
+    assertThat(handleHolder.getHandle("writeHandle")).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun setHandle_handelAlreadySet_asserts() {
+    val handleHolder = DummyHandleHolder()
+    handleHolder.setHandle("writeHandle", mock())
+
+    assertFailsWith<IllegalArgumentException> {
+      handleHolder.setHandle("writeHandle", mock())
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "TestParticle.writeHandle has already been initialized."
+      )
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameNotInSpec_throwsNoElemException() {
+    val handleHolder = DummyHandleHolder()
+
+    assertFailsWith<NoSuchElementException> {
+      handleHolder.setHandle("handleButNotEntities", mock())
+    }.also {
+      assertThat(it.message).isEqualTo(
+        "Particle TestParticle does not have a handle with name handleButNotEntities."
+      )
+    }
+  }
+
+  @Test
+  fun setHandle_handelNameInSpec_success() {
+    val handleHolder = DummyHandleHolder()
+    val mock = mock<Handle>()
+
+    handleHolder.setHandle("writeHandle", mock)
+
+    assertThat(handleHolder.writeHandle).isSameInstanceAs(mock)
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap() {
+    val handleHolder = DummyHandleHolder()
+
+    assertThat(handleHolder.isEmpty()).isTrue()
+
+    handleHolder.setHandle("readHandle", mock())
+
+    assertThat(handleHolder.isEmpty()).isFalse()
+  }
+
+  @Test
+  fun createForeignReference_noHandles_asserts() = runBlockingTest {
+    val handleHolder = DummyHandleHolder()
+
+    assertFailsWith<IllegalStateException> {
+      handleHolder.createForeignReference<EntityBase>(mock(), "someId")
+    }
+  }
+
+  @Test
+  fun createForeignReference_someHandles_success() = runBlockingTest {
+    val handleHolder = DummyHandleHolder()
+    val mock = mock<Handle>()
+    handleHolder.setHandle("readHandle", mock)
+
+    handleHolder.readHandle.createForeignReference<EntityBase>(mock(), "someId")
+
+    verify(mock, times(1)).createForeignReference<EntityBase>(any(), any())
+  }
+
+  companion object {
+    val DUMMY_DISPATCHER_1 = Dispatchers.Default
+    val DUMMY_DISPATCHER_2 = Dispatchers.Main
+
+    class DummyHandleHolder : HandleHolderBase(
+      "TestParticle",
+      mapOf(
+        "readHandle" to setOf(mock()),
+        "writeHandle" to setOf(mock()),
+        "entityButNotHandles" to setOf(mock())
+      )
+    ) {
+      val readHandle by handles
+      val writeHandle by handles
+      val handleButNotEntities by handles
```

Comment:
Related to the comment in the first two tests: this property-based structure is related to the generated sublcasses, not HandleHolderBase itself. So there's no need to create subclasses with properties like this, since we have test coverage for that elsewhere.